File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-1.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        set<int> rows, cols;
        int m = matrix.size();
        int n = matrix[0].size();

        for (int i=0; i < matrix.size(); i++) {
            for (int j=0; j<matrix[i].size(); j++) {
                if (matrix[i][j] == 0) {
                    rows.insert(i);
                    cols.insert(j);
                }
            }
        }

        for (auto it = rows.begin(); it != rows.end(); it++) {
            int x = *it;

            for (int i=0; i<n; i++) {
                matrix[x][i] = 0;
            }
        }

        for (auto it = cols.begin(); it != cols.end(); it++) {
            int y = *it;

            for (int i=0; i<m; i++) {
                matrix[i][y] = 0;
            }
        }
    }
};

File: Data Structures & Algorithms/valid-sudoku/submission-1.js
class Solution {
    /**
     * @param {character[][]} board
     * @return {boolean}
     */
    isValidSudoku(board) {
        // valid rows
        // valid cols
        //valid 3x3

        for (let i=0; i<9; i++) {
            if (!this.isValidRow(board, i)) {
                console.log('no valid row', i);
                return false;
            }
        }

        for (let i=0 ;i<9; i++) {
            if (!this.isValidCol(board, i)) {
                console.log('not valid col', i);
                return false;
            }
        }

        for (let i=0;i<3;i++) {
            for (let j=0; j<3; j++) {
                if (!this.isValidMiniSquare(board, 3*i, 3*j)) {
                    console.log('not valid mini sq', 3*i , 3*j);
                    return false;
                }
            }
        }

        return true;
    }

    isValidRow(board, row) {
        const map = new Map();
        for (let i=0;i < 9; i++) {
            if (board[row][i] == '.') continue;

            if (map.has(board[row][i])) {
                return false;
            } else {
                map.set(board[row][i], true);
            }
        }

        return true;
    }

    isValidCol(board, col) {
        const map = new Map();
        for (let i=0;i < 9; i++) {
            if (board[i][col] == '.') continue; 
            if (map.has(board[i][col])) {
                return false;
            } else {
                map.set(board[i][col], true);
            }
        }

        return true;
    }

    isValidMiniSquare(board, x, y) {
        const map = new Map();
        for (let i=x; i<x+3; i++) {
            for (let j=y; j<y+3; j++) {
                if (board[i][j] == '.') continue; 
                
                if (map.has(board[i][j])) {
                    return false;
                } else {
                    map.set(board[i][j], true);
                }
            }
        }

        return true;
    }
}
